src/client/datascience/jupyter/debuggerVariables.ts

```diff
@@ -452,6 +415,65 @@ export class DebuggerVariables extends DebugLocationTracker
         const activeNotebook = this.vscNotebook.activeNotebookEditor;
         return !!activeNotebook && this.debuggingManager.isDebugging(activeNotebook.document);
     }
+
+    // This handles all the debug session calls, variable handling, and refresh calls needed for notebook debugging
+    private async handleNotebookVariables(stoppedMessage: DebugProtocol.StoppedEvent): Promise<void> {
+        const doc = this.vscNotebook.activeNotebookEditor?.document;
+        const threadId = stoppedMessage.body.threadId;
+
+        if (doc) {
+            const session = await this.debuggingManager.getDebugSession(doc);
+            if (session) {
+                // Call stack trace
+                const stResponse = await session.customRequest('stackTrace', {
+                    threadId,
+                    startFrame: 0,
+                    levels: 1
+                });
+
+                //  Call scopes
+                const sf = stResponse.stackFrames[0];
+                const mode = this.debuggingManager.getDebugMode(doc);
+                // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                let scopesResponse: any | undefined;
+
+                if (mode === KernelDebugMode.RunByLine) {
+                    // Only call scopes (and variables) if we are stopped on the cell we are executing
+                    const cell = this.debuggingManager.getDebugCell(doc);
+                    if (sf.source && cell && sf.source.path === cell.document.uri.toString()) {
+                        scopesResponse = await session.customRequest('scopes', { frameId: sf.id });
+                    }
+                } else {
+                    // Only call scopes (and variables) if we are stopped on the notebook we are executing
+                    const docURI = path.basename(doc.uri.toString());
+                    if (sf.source && sf.source.path && sf.source.path.includes(docURI)) {
+                        scopesResponse = await session.customRequest('scopes', { frameId: sf.id });
+                    }
+                }
+
+                // Call variables
+                if (scopesResponse) {
+                    // Keep track of variablesReference because "hover" requests also try to update variables
+                    const newVariablesReference = scopesResponse.scopes[0].variablesReference;
```

Comment:
Are you sure you only want to update the first scope?

Reply:
no hahah